src/components/PivotTable/PivotTableRenderer/PivotTableBuilder.tsx

```diff
@@ -270,145 +220,254 @@ function getVertical<T extends object>({
     cellPositions.add(gridArea.toString())
   }
 
-  /**
-   * Create horizontal header
-   */
-  divs.push(
-    ...keys.map((k, i) => {
-      const gridArea = new GridArea(i + 1, columnHeaderSpace, i + 2, columnHeaderSpace + 1)
-      return (
-        <PivotTableCell types={new Set([PivotTableCellType.HEADER])} key={gridArea.toString()} gridArea={gridArea}>
-          {keysMapping.get(k).keyName}
-        </PivotTableCell>
-      )
-    })
-  )
-
-  /**
-   * Populate totals
-   */
-  let totaisGridArea: GridArea
-  let dataValueGridArea: GridArea
   if (mixedTable) {
-    /**
-     * Populate horizontal totals
-     */
-    const totalRowNumber = mixedTable.totalRowNumber
-    totaisGridArea = new GridArea(1, maxColumnEnd + 1, keys.length + 2, maxColumnEnd + 2)
-    dataValueGridArea = new GridArea(totalRowNumber, maxColumnEnd + 1, totalRowNumber + 1, maxColumnEnd + 2)
+    buildMixedTableTotals<T>(
+      maxColumnEnd,
+      keys,
+      mixedTable.rowTotalValues,
+      mixedTableColumnTotals,
+      mixedTable.totalRowNumber,
+      divs,
+      cellPositions,
+      mixedTableStartRowCache,
+      columnHeaderSpace,
+      data.nodeValue
+    )
+  } else {
+    buildVerticalTableTotals(maxRowEnd, maxColumnEnd, divs, data.nodeValue)
+  }
+  return divs
+}
 
-    mixedTableColumnTotals.forEach((value, key) => {
-      const gridArea = new GridArea(totalRowNumber, key, totalRowNumber + 1, key + 1)
-      divs.push(
-        <PivotTableCell
-          types={new Set([PivotTableCellType.TOTAL, PivotTableCellType.VALUE])}
-          isEndRow
-          key={gridArea.toString()}
-          gridArea={gridArea}
-        >
-          {value}
-        </PivotTableCell>
-      )
-      cellPositions.add(gridArea.toString())
-    })
+function buildHorizontalTableHeader<T extends object>(
+  keys: (keyof T)[],
+  rowHeaderSpace: number,
+  divs: ReactElement[],
+  keysMapping: KeyMap<T>
+) {
```

Comment:
It is possible, but because the `divs` array can get too big and might be a performance bottleneck, to `concat` the results of every function to it would create a lot of clones of `divs`.